38 document additional knowledge interesting aspects (#39)
cloudogu/dogu-development-docs: docs/additional/interesting_aspects_en.md

Using this approach, it is possible to customize all files of the container.

### Bash scripts

The creation of bash scripts in a dogu is necessary to control various processes.  
This includes starting and upgrading the dogu, as well as creating a service account.

In general, all scripts should conform to the following conventions:

#### bash-strict-mode

Bash scripts should always use [bash-strict-mode][strict-mode].  
This prevents further execution of the script in the event of a misbehavior.
In general, it makes sense to catch all potential errors in the startup script and report them with a separate error method.  
This could look like this:

```shell
# This function prints an error to the console and waits 5 minutes before exiting the process.
function exitOnErrorWithMessage() {
  message=${1}
  echo -e "ERROR: ${message}. Exiting in 300 seconds."
  sleep 300
  exit 1
}
```

Such a method makes the whole script more readable and slows down the restart loop of the Dogu as well as the mass of generated logs.

[strict-mode]: http://redsymbol.net/articles/unofficial-bash-strict-mode/

#### Line-Endings

> Unix-based line endings should be used for all scripts at all times.#### Use doguctl

Reviewer: Translation fail?
```suggestion
Unix-based line endings should be used for all scripts at all times.
```

This includes aspects like:
- Checking permission for scripts (startup.sh), volumes.
- Checking the UID & GID for files (startup.sh, resources, volumes, write-dependent paths)
- Accessibility of the software: TCP health checks

For pipeline integration, we recommend using the `ecoSystem.verify()` method from the [dogu-build-lib][dogu-build-lib].  
An example call looks like the following in a Jenkinsfile:

```groovy
stage('Verify') {
    ecoSystem.verify("/dogu")
}
```

This internally calls the `cesapp verify` command, which runs the goss tests.

[goss]: https://github.com/goss-org/goss
[dogu-build-lib]: https://github.com/cloudogu/dogu-build-lib

## Documentation

Every tool should have a reasonable documentation about used features, which supports usage or support.
To allow easy integration into our existing documentation infrastructure, the documentation should be placed in a `/docs` folder located in the repository root.

By default, the docs include the following:
- Features of the Dogu
- Installation of the Dogu
- Configuration of the Dogu
- Development on the Dogu
- System requirements of the Dogu (nice to have)
- Readme

> - Changelog

Reviewer: This reads as if the Readme belongs to docs/ which is not our default with the Readme.md being the documentation starting page in the project root

The same goes for the changlog

### License

The license for using the Dogu should be clearly stated.

> This can be specified in the Readme, for example.

Reviewer: It would also help to provide a full license file if one exists, f.i. in case of a publicly available repository (it does not have to be a FOSS repository)

### Readme

> A Readme is useful both for your own developments and in a partner context with external developers.  

Reviewer: We should start with explaining what a Readme file makes.

```suggestion
A Readme is a file consisting of diverse and often importan information about the given software. It is useful both for your own developments and in a partner context with external developers.  
```

In the case of public repositories, however, a higher standard should be applied to a Readme.

> A Readme file is located in the repository root and concisely describes notes on the general purpose of the Dogu.
> This can be in the form of a quickstart guide, for example, describing how to set up a local development environment.
> It should also include a reference to the documentation folder and can also refer to specific pages of the documentation.
> Finally, it is recommended to specify required resources for use and a person responsible for the Dogu.

Reviewer: This sentences mixes two things which should be separated:

1. describing the purpose of the dogu (aimed at dogu users)
2. a development quickstart (aimed at dogu developers)

I suggest to make the possible uses of a Readme more plastic, like so:

```suggestion
A Readme file is located in the repository root and concisely describes notes on the general purpose of the Dogu. For users this can be some sorts of quickstart guide of how to use the Dogu, or even a feature list. For Dogu developers the readme may provide insights of how collaboration is possible or which copyright license was chosen. Additional docs can be linked to provide further information for both users and developers.

Administrators may want to find information about the resource consumption of this dogu. Finally, the readme should clarify who is responsible for this Dogu.
```

### Changelog

Changelogs are written for people who want to understand what's new in releases.

The format of the changelog should follow **[keep a changelog](https://keepachangelog.com)**.

New changes are always added to the top section `## [Unreleased]`.  
When a new release is made, these changes move to a new section containing the corresponding version number.

For all changes, references to the operations of the issue tracker used should be added.  
In addition, if more complex procedures are involved, a reference to documentation is useful.

### System requirements

The system requirements of a Dogu (CPU, RAM, storage) can either be recorded in the Dogu's documentation
or should be otherwise communicated to Cloudogu directly (e.g. in case of a private repository).

## Release of a Dogus

### Cloudogu account and permissions

> First, you need to create an account on https://account.cloudogu.com.  

Reviewer: This paragraph should mention that this step is only necessary once per devleoper account.
